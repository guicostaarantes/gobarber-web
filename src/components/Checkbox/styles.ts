import styled from 'styled-components';

import { CONSTANTS } from '../../styles/constants';

export const Container = styled.div`
  padding: 16px;
  width: 100%;
  display: flex;
  align-items: center;
  margin: 8px;
`;

export const Input = styled.input`
  appearance: none;
  width: 38px;
  height: 21px;
  border-radius: 11px;
  outline: none;
  display: inline-block;
  vertical-align: top;
  position: relative;
  margin: 0;
  cursor: pointer;
  border: 1px solid var(--bc, ${CONSTANTS.grayColor6});
  background: var(--b, ${CONSTANTS.grayColor7});
  transition: background 0.3s, border-color 0.3s, box-shadow 0.3s;

  &:after {
    content: '';
    display: block;
    position: absolute;
    left: 2px;
    top: 2px;
    border-radius: 50%;
    width: 15px;
    height: 15px;
    background: var(--ab, ${CONSTANTS.primaryColor});
    transform: translateX(var(--x, 0));
    transition: transform 0.3s, opacity 0.3s;
  }

  &:checked {
    --b: ${CONSTANTS.primaryDarkerColor};
    --bc: ${CONSTANTS.primaryColor};
    --ab: ${CONSTANTS.primaryColor};
    --x: 17px;

    & + label {
      & .active {
        display: block;
      }
      & .inactive {
        display: none;
      }
    }
  }

  &:disabled {
    --ab: ${CONSTANTS.grayColor5};
    cursor: not-allowed;
    opacity: 0.9;

    & + label {
      cursor: not-allowed;
    }
  }

  &:hover {
    &:not(:checked) {
      &:not(:disabled) {
        --bc: ${CONSTANTS.primaryDarkColor};
      }
    }
  }

  &:focus {
    box-shadow: 0 0 0 ${CONSTANTS.primaryColor};
  }

  & + label {
    color: ${CONSTANTS.grayColor7};
    display: inline-block;
    vertical-align: top;
    cursor: pointer;
    margin-left: 8px;
    & .active {
      display: none;
    }
  }
`;

export const Label = styled.label``;
