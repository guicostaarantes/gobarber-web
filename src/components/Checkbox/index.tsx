import React, {
  InputHTMLAttributes,
  useRef,
  useEffect,
  useState,
  useCallback,
} from 'react';

import { useField } from '@unform/core';

import { IconBaseProps } from 'react-icons';

import { Container, Input, Label } from './styles';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name: string;
  label: string;
  uncheckedLabel?: string;
  icon?: React.ComponentType<IconBaseProps>;
}

const Checkbox: React.FC<InputProps> = ({
  name,
  label,
  uncheckedLabel,
  ...otherProps
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const { fieldName, registerField } = useField(name);

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: inputRef.current,
      path: 'checked',
    });
  }, [fieldName, registerField]);

  return (
    <Container>
      <Input
        id={fieldName}
        ref={inputRef}
        type="checkbox"
        data-testid="checkbox"
        {...otherProps}
      />
      <Label htmlFor={fieldName} data-testid="checkbox-label">
        <span className="active">{label}</span>
        <span className="inactive">{uncheckedLabel || label}</span>
      </Label>
    </Container>
  );
};

export default Checkbox;
