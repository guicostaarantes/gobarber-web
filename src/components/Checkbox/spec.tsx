import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Form } from '@unform/web';
import { FormHandles } from '@unform/core';
import Checkbox from './index';

describe('Checkbox unit test', () => {
  it('should render correctly', () => {
    const fn = jest.fn();
    const formRef = React.createRef<FormHandles>();
    const { container } = render(
      <Form ref={formRef} onSubmit={fn}>
        <Checkbox
          name="newsletter"
          label="I would like to receive news"
          uncheckedLabel="I would NOT like to receive news"
        />
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should change when typing', async () => {
    const fn = jest.fn();
    const formRef = React.createRef<FormHandles>();
    const { getByTestId } = render(
      <Form ref={formRef} onSubmit={fn}>
        <Checkbox
          name="newsletter"
          label="I would like to receive news"
          uncheckedLabel="I would NOT like to receive news"
        />
      </Form>,
    );
    const input = getByTestId('checkbox');
    fireEvent.change(input, { target: { checked: true } });
    expect(formRef.current?.getData()).toEqual({ newsletter: true });
    fireEvent.change(input, { target: { checked: false } });
    expect(formRef.current?.getData()).toEqual({ newsletter: false });
  });
});
