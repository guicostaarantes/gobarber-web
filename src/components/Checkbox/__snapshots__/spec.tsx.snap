// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox unit test should render correctly 1`] = `
.c0 {
  padding: 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 8px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 38px;
  height: 21px;
  border-radius: 11px;
  outline: none;
  display: inline-block;
  vertical-align: top;
  position: relative;
  margin: 0;
  cursor: pointer;
  border: 1px solid var(--bc,#999591);
  background: var(--b,#dddddd);
  -webkit-transition: background 0.3s,border-color 0.3s,box-shadow 0.3s;
  transition: background 0.3s,border-color 0.3s,box-shadow 0.3s;
}

.c1:after {
  content: '';
  display: block;
  position: absolute;
  left: 2px;
  top: 2px;
  border-radius: 50%;
  width: 15px;
  height: 15px;
  background: var(--ab,#ff9000);
  -webkit-transform: translateX(var(--x,0));
  -ms-transform: translateX(var(--x,0));
  transform: translateX(var(--x,0));
  -webkit-transition: -webkit-transform 0.3s,opacity 0.3s;
  -webkit-transition: transform 0.3s,opacity 0.3s;
  transition: transform 0.3s,opacity 0.3s;
}

.c1:checked {
  --b: #a95f00;
  --bc: #ff9000;
  --ab: #ff9000;
  --x: 17px;
}

.c1:checked + label .active {
  display: block;
}

.c1:checked + label .inactive {
  display: none;
}

.c1:disabled {
  --ab: #666360;
  cursor: not-allowed;
  opacity: 0.9;
}

.c1:disabled + label {
  cursor: not-allowed;
}

.c1:hover:not(:checked):not(:disabled) {
  --bc: #cc7300;
}

.c1:focus {
  box-shadow: 0 0 0 #ff9000;
}

.c1 + label {
  color: #dddddd;
  display: inline-block;
  vertical-align: top;
  cursor: pointer;
  margin-left: 8px;
}

.c1 + label .active {
  display: none;
}

<div>
  <form>
    <div
      class="c0"
    >
      <input
        class="c1"
        data-testid="checkbox"
        id="newsletter"
        type="checkbox"
      />
      <label
        class=""
        for="newsletter"
      >
        <span
          class="active"
        >
          I would like to receive news
        </span>
        <span
          class="inactive"
        >
          I would NOT like to receive news
        </span>
      </label>
    </div>
  </form>
</div>
`;
